lib/runtime-core/src/state.rs

```diff
@@ -0,0 +1,843 @@
+}
+
+#[derive(Clone, Debug)]
+pub struct MachineState {
+    pub stack_values: Vec<MachineValue>,
+    pub register_values: Vec<MachineValue>,
+
+    pub wasm_stack: Vec<WasmAbstractValue>,
+    pub wasm_stack_private_depth: usize,
+
+    pub wasm_inst_offset: usize,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct MachineStateDiff {
+    pub last: Option<usize>,
+    pub stack_push: Vec<MachineValue>,
+    pub stack_pop: usize,
+    pub reg_diff: Vec<(RegisterIndex, MachineValue)>,
+
+    pub wasm_stack_push: Vec<WasmAbstractValue>,
+    pub wasm_stack_pop: usize,
+    pub wasm_stack_private_depth: usize, // absolute value; not a diff.
+
+    pub wasm_inst_offset: usize, // absolute value; not a diff.
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum MachineValue {
+    Undefined,
+    Vmctx,
+    PreserveRegister(RegisterIndex),
+    CopyStackBPRelative(i32), // relative to Base Pointer, in byte offset
+    ExplicitShadow,           // indicates that all values above this are above the shadow region
+    WasmStack(usize),
+    WasmLocal(usize),
+}
+
+#[derive(Clone, Debug)]
+pub struct FunctionStateMap {
+    pub initial: MachineState,
+    pub local_function_id: usize,
+    pub locals: Vec<WasmAbstractValue>,
+    pub shadow_size: usize, // for single-pass backend, 32 bytes on x86-64
+    pub diffs: Vec<MachineStateDiff>,
+    pub wasm_function_header_target_offset: Option<SuspendOffset>,
+    pub wasm_offset_to_target_offset: BTreeMap<usize, SuspendOffset>,
+    pub loop_offsets: BTreeMap<usize, OffsetInfo>, /* suspend_offset -> info */
+    pub call_offsets: BTreeMap<usize, OffsetInfo>, /* suspend_offset -> info */
+    pub trappable_offsets: BTreeMap<usize, OffsetInfo>, /* suspend_offset -> info */
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum SuspendOffset {
+    Loop(usize),
+    Call(usize),
+    Trappable(usize),
+}
+
+#[derive(Clone, Debug)]
+pub struct OffsetInfo {
+    pub diff_id: usize,
+    pub activate_offset: usize,
+}
+
+#[derive(Clone, Debug)]
+pub struct ModuleStateMap {
+    pub local_functions: BTreeMap<usize, FunctionStateMap>,
+    pub total_size: usize,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct WasmFunctionStateDump {
+    pub local_function_id: usize,
+    pub wasm_inst_offset: usize,
+    pub stack: Vec<Option<u64>>,
+    pub locals: Vec<Option<u64>>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ExecutionStateImage {
+    pub frames: Vec<WasmFunctionStateDump>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct InstanceImage {
+    pub memory: Option<Vec<u8>>,
+    pub globals: Vec<u64>,
+    pub execution_state: ExecutionStateImage,
+}
+
+impl ModuleStateMap {
+    fn lookup_call_ip(&self, ip: usize, base: usize) -> Option<(&FunctionStateMap, MachineState)> {
+        if ip < base || ip - base >= self.total_size {
+            None
+        } else {
+            //println!("lookup ip: {} in {:?}", ip - base, self.local_functions);
```

Comment:
Remove commented out code or make it `debug`.

Reply:
or make it `trace!` if it's very verbose 😸